Question: Cannot compile standard OpenSSL code using OQS-OpenSSL (already configured and installed)

I have compiled OQS-OpenSSL on CentOS 7.7 using gcc 7.3.1 by following the instructions (First get OQS-OpenSSL, then build and install liboqs at the root of OQS-OpenSSL and finally build OQS-OpenSSL). 

Everything seems to work during the build process of OQS-OpenSSL, I've installed OQS-OpenSSL into "/usr/local/1.1.1d-OQS-OpenSSL_1_1_1-stable_snapshot_2019-10-0.2.0"

**Issue:**
The issue I get is when I try to compile code using standard OpenSSL is it complains that "oqs/oqs.h" couldn't be found. Indeed "oqs/oqs.h" is not installed when I issue a "make install" inside the OQS-OpenSSL directory.  Also the instructions doesn't tell to separately install liboqs outside of OQS-OpenSSL.

**Here's my build process:**
```
cd /root/build/openssl/openssl-OpenSSL_1_1_1d && make clean ; ./Configure shared linux-x86_64 -lm --prefix="/usr/local/openssl-1.1.1d" && make -j 4 && make install
cd /root/build/liboqs/liboqs-0.2.0/ && autoreconf -i
cd /root/build/liboqs/liboqs-0.2.0/ && make clean ; ./configure --enable-shared --prefix="/root/build/openssloqs/openssl-OQS-OpenSSL_1_1_1-stable-snapshot-2019-10//oqs" --with-openssl="/usr/local/openssl-1.1.1d" "--disable-kem-bike" && make -j 4 && make install
cd /root/build/openssloqs/openssl-OQS-OpenSSL_1_1_1-stable-snapshot-2019-10/ && make clean ; ./Configure --prefix="/usr/local/1.1.1d-OQS-OpenSSL_1_1_1-stable_snapshot_2019-10-0.2.0" shared linux-x86_64 -lm && make -j 4 && make install
```
Now, when I try to build code that uses standard OpenSSL with OQS-OpenSSL I get this error (evp.h)

**Error:**
```
/usr/local/1.1.1d-OQS-OpenSSL_1_1_1-stable_snapshot_2019-10-0.2.0/include/openssl/evp.h:18:11: fatal error: oqs/oqs.h: No such file or directory
include <oqs/oqs.h>
       ^~~~~~~~~~~
compilation terminated.
make: *** [: OutboundTransferMgr.o] Error 1
[ CHROOT ] bash-4.2$>
```


==> How should I use the OpenSSL fork when building software that uses OpenSSL?
I have added 
```
"-I/root/build/openssloqs/openssl-OQS-OpenSSL_1_1_1-stable-snapshot-2019-10/oqs/include"
```
to the Makefile so that "oqs/oqs.h" could resolve but I get other strange errors following this.

Thanks for your work here!  I appreciate it :)

Sylvain

Hi Sylvain,

 

   as you seem to want to build a ‘normal openssl enabled app’, did you follow all guidance posted here: https://github.com/open-quantum-safe/openssl/wiki/Running-TLS-1.3-in-nginx#download-and-build-nginx ?

 

In addition, two observations :
I’m not sure this typo (double ‘//’) in your config is “somewhat” OK or must be corrected (I’d fix it anyway): root/build/openssloqs/openssl-OQS-OpenSSL_1_1_1-stable-snapshot-2019-10//oqs
Make install within oqs-openssl does not seem to _install_ OQS dependencies, too: I could look into automating this if no-one else can provide a pointer as to how to facilitate that (at first glance it seems to be a pkg-config issue). Until that’s available, feel free to copy oqs libraries and includes into the proper locations for your further build to succeed. In your case, I’d recommend the following (no guarantees, this is a reading-your-mail-only exercise): 
Libraries: cp /root/build/openssloqs/openssl-OQS-OpenSSL_1_1_1-stable-snapshot-2019-10/oqs/libs/* /usr/local/1.1.1d-OQS-OpenSSL_1_1_1-stable_snapshot_2019-10-0.2.0/lib (and don’t forget to add oqs libs to your app’s build process as described in the link above)
Includes: cp -R /root/build/openssloqs/openssl-OQS-OpenSSL_1_1_1-stable-snapshot-2019-10/oqs/include/oqs /usr/local/1.1.1d-OQS-OpenSSL_1_1_1-stable_snapshot_2019-10-0.2.0/include
à With this, you should be “getting away” with pointing to /root/build/openssloqs/openssl-OQS-OpenSSL_1_1_1-stable-snapshot-2019-10 as the ‘non-standard OpenSSL location’ that you typically have to define when building an openssl-consuming app (with openssl in a non-standard-location)

 

Hope this helps,

 

--Michael

 

From: syrtap <notifications@github.com>
Reply to: open-quantum-safe/openssl <reply@reply.github.com>
Date: Tuesday, 29 October 2019 at 19:12
To: open-quantum-safe/openssl <openssl@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: [EXTERNAL] [open-quantum-safe/openssl] Question: Cannot compile standard OpenSSL code using OQS-OpenSSL (already configured and installed) (#128)

 

I have compiled OQS-OpenSSL on CentOS 7.7 using gcc 7.3.1 by following the instructions (First get OQS-OpenSSL, then build and install liboqs at the root of OQS-OpenSSL and finally build OQS-OpenSSL).

Everything seems to work during the build process of OQS-OpenSSL, I've installed OQS-OpenSSL into "/usr/local/1.1.1d-OQS-OpenSSL_1_1_1-stable_snapshot_2019-10-0.2.0"

Issue:
The issue I get is when I try to compile code using standard OpenSSL is it complains that "oqs/oqs.h" couldn't be found. Indeed "oqs/oqs.h" is not installed when I issue a "make install" inside the OQS-OpenSSL directory. Also the instructions doesn't tell to separately install liboqs outside of OQS-OpenSSL.

Here's my build process:
cd /root/build/openssl/openssl-OpenSSL_1_1_1d && make clean ; ./Configure shared linux-x86_64 -lm --prefix="/usr/local/openssl-1.1.1d" && make -j 4 && make install
 
cd /root/build/liboqs/liboqs-0.2.0/ && autoreconf -i
 
cd /root/build/liboqs/liboqs-0.2.0/ && make clean ; ./configure --enable-shared --prefix="/root/build/openssloqs/openssl-OQS-OpenSSL_1_1_1-stable-snapshot-2019-10//oqs" --with-openssl="/usr/local/openssl-1.1.1d" "--disable-kem-bike" && make -j 4 && make install
 
cd /root/build/openssloqs/openssl-OQS-OpenSSL_1_1_1-stable-snapshot-2019-10/ && make clean ; ./Configure --prefix="/usr/local/1.1.1d-OQS-OpenSSL_1_1_1-stable_snapshot_2019-10-0.2.0" shared linux-x86_64 -lm && make -j 4 && make install
 
Now, when I try to build code that uses standard OpenSSL with OQS-OpenSSL I get this error (evp.h)

Error:
/usr/local/1.1.1d-OQS-OpenSSL_1_1_1-stable_snapshot_2019-10-0.2.0/include/openssl/evp.h:18:11: fatal error: oqs/oqs.h: No such file or directory
 
include <oqs/oqs.h>
 
       ^~~~~~~~~~~
 
compilation terminated.
 
make: *** [: OutboundTransferMgr.o] Error 1
 
[ CHROOT ] bash-4.2$>
 
==> How should I use the OpenSSL fork when building software that uses OpenSSL?
I have added
"-I/root/build/openssloqs/openssl-OQS-OpenSSL_1_1_1-stable-snapshot-2019-10/oqs/include"
 
to the Makefile so that "oqs/oqs.h" could resolve but I get other strange errors following this.

Thanks for your work here! I appreciate it :)

Sylvain

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or unsubscribe.

--- a/Configurations/unix-Makefile.tmpl
+++ b/Configurations/unix-Makefile.tmpl
@@ -527,6 +527,7 @@ clean: libclean
 	$(RM) openssl.pc libcrypto.pc libssl.pc liboqs.pc
 	-$(RM) `find . -type l \! -name '.*' -print`
 	$(RM) $(TARFILE)
+	-$(RM) -rf oqs
 
 distclean: clean
 	$(RM) configdata.pm
@@ -678,8 +679,10 @@ install_dev: install_runtime_libs
 	@$(ECHO) "install liboqs.pc -> $(DESTDIR)$(libdir)/pkgconfig/liboqs.pc"
 	@cp liboqs.pc $(DESTDIR)$(libdir)/pkgconfig
 	@install oqs/lib/liboqs.a $(DESTDIR)$(libdir)
+ifneq (,$(wildcard oqs/lib/liboqs.so.0.0.0))
 	@install oqs/lib/liboqs.so.0.0.0 $(DESTDIR)$(libdir)
 	@cd $(DESTDIR)$(libdir) && ln -sf liboqs.so.0.0.0 liboqs.so && cd -
+endif
 	@chmod 644 $(DESTDIR)$(libdir)/pkgconfig/liboqs.pc
 
 uninstall_dev: uninstall_runtime_libs

--- a/Configurations/unix-Makefile.tmpl
+++ b/Configurations/unix-Makefile.tmpl
@@ -527,11 +527,11 @@ clean: libclean
 	$(RM) openssl.pc libcrypto.pc libssl.pc liboqs.pc
 	-$(RM) `find . -type l \! -name '.*' -print`
 	$(RM) $(TARFILE)
-	-$(RM) -rf oqs
 
 distclean: clean
 	$(RM) configdata.pm
 	$(RM) Makefile
+	-$(RM) -rf oqs
 
 # We check if any depfile is newer than Makefile and decide to
 # concatenate only if that is true.

--- a/Configurations/unix-Makefile.tmpl
+++ b/Configurations/unix-Makefile.tmpl
@@ -453,7 +453,7 @@ LANG=C
 {- dependmagic('build_programs'); -}: build_programs_nodep
 
 build_generated: $(GENERATED_MANDATORY)
-build_libs_nodep: libcrypto.pc libssl.pc openssl.pc
+build_libs_nodep: libcrypto.pc libssl.pc openssl.pc liboqs.pc
 build_engines_nodep: $(ENGINES)
 build_programs_nodep: $(PROGRAMS) $(SCRIPTS)
 
@@ -524,7 +524,7 @@ clean: libclean
 	$(RM) core
 	$(RM) tags TAGS doc-nits
 	$(RM) -r test/test-runs
-	$(RM) openssl.pc libcrypto.pc libssl.pc
+	$(RM) openssl.pc libcrypto.pc libssl.pc liboqs.pc
 	-$(RM) `find . -type l \! -name '.*' -print`
 	$(RM) $(TARFILE)
 
@@ -613,6 +613,13 @@ install_dev: install_runtime_libs
 		cp $$i $(DESTDIR)$(INSTALLTOP)/include/openssl/$$fn; \
 		chmod 644 $(DESTDIR)$(INSTALLTOP)/include/openssl/$$fn; \
 	done
+	@$(PERL) $(SRCDIR)/util/mkdir-p.pl $(DESTDIR)$(INSTALLTOP)/include/oqs
+	@set -e; for i in $(SRCDIR)/oqs/include/oqs/*.h; do \
+		fn=`basename $$i`; \
+		$(ECHO) "install $$i -> $(DESTDIR)$(INSTALLTOP)/include/oqs/$$fn"; \
+		cp $$i $(DESTDIR)$(INSTALLTOP)/include/oqs/$$fn; \
+		chmod 644 $(DESTDIR)$(INSTALLTOP)/include/oqs/$$fn; \
+	done
 	@$(PERL) $(SRCDIR)/util/mkdir-p.pl $(DESTDIR)$(libdir)
 	@set -e; for l in $(INSTALL_LIBS); do \
 		fn=`basename $$l`; \
@@ -668,6 +675,12 @@ install_dev: install_runtime_libs
 	@$(ECHO) "install openssl.pc -> $(DESTDIR)$(libdir)/pkgconfig/openssl.pc"
 	@cp openssl.pc $(DESTDIR)$(libdir)/pkgconfig
 	@chmod 644 $(DESTDIR)$(libdir)/pkgconfig/openssl.pc
+	@$(ECHO) "install liboqs.pc -> $(DESTDIR)$(libdir)/pkgconfig/liboqs.pc"
+	@cp liboqs.pc $(DESTDIR)$(libdir)/pkgconfig
+	@install oqs/lib/liboqs.a $(DESTDIR)$(libdir)
+	@install oqs/lib/liboqs.so.0.0.0 $(DESTDIR)$(libdir)
+	@cd $(DESTDIR)$(libdir) && ln -sf liboqs.so.0.0.0 liboqs.so && cd -
+	@chmod 644 $(DESTDIR)$(libdir)/pkgconfig/liboqs.pc
 
 uninstall_dev: uninstall_runtime_libs
 	@$(ECHO) "*** Uninstalling development files"
@@ -710,6 +723,7 @@ uninstall_dev: uninstall_runtime_libs
 	$(RM) $(DESTDIR)$(libdir)/pkgconfig/libcrypto.pc
 	$(RM) $(DESTDIR)$(libdir)/pkgconfig/libssl.pc
 	$(RM) $(DESTDIR)$(libdir)/pkgconfig/openssl.pc
+	$(RM) $(DESTDIR)$(libdir)/pkgconfig/liboqs.pc
 	-$(RMDIR) $(DESTDIR)$(libdir)/pkgconfig
 	-$(RMDIR) $(DESTDIR)$(libdir)
 
@@ -989,6 +1003,23 @@ libssl.pc:
 	    echo 'Libs: -L$${libdir} -lssl'; \
 	    echo 'Cflags: -I$${includedir}' ) > libssl.pc
 
+liboqs.pc:
+	@ ( echo 'prefix=$(INSTALLTOP)'; \
+	    echo 'exec_prefix=$${prefix}'; \
+	    if [ -n "$(LIBDIR)" ]; then \
+	        echo 'libdir=$${exec_prefix}/$(LIBDIR)'; \
+	    else \
+	        echo 'libdir=$(libdir)'; \
+	    fi; \
+	    echo 'includedir=$${prefix}/include'; \
+	    echo ''; \
+	    echo 'Name: liboqs'; \
+	    echo 'Description: Open Quantum Safe crypto libraries'; \
+	    echo 'Version: '$(VERSION); \
+	    echo 'Requires.private: libssl libcrypto'; \
+	    echo 'Libs: -L$${libdir} -loqs'; \
+	    echo 'Cflags: -I$${includedir}' ) > liboqs.pc
+
 openssl.pc:
 	@ ( echo 'prefix=$(INSTALLTOP)'; \
 	    echo 'exec_prefix=$${prefix}'; \
@@ -1002,7 +1033,7 @@ openssl.pc:
 	    echo 'Name: OpenSSL'; \
 	    echo 'Description: Secure Sockets Layer and cryptography libraries and tools'; \
 	    echo 'Version: '$(VERSION); \
-	    echo 'Requires: libssl libcrypto' ) > openssl.pc
+	    echo 'Requires: libssl libcrypto liboqs' ) > openssl.pc
 
 configdata.pm: $(SRCDIR)/Configure $(SRCDIR)/config {- join(" ", @{$config{build_file_templates}}, @{$config{build_infos}}, @{$config{conf_files}}) -}
 	@echo "Detected changed: $?"

